fix update and delete tenant user

critBus opened the pull request:
### Update:

The update was modified so that it was only possible to edit the value of the is_default field

In the case of editing, if you try to change the default value to false and there is no other entity available, then a bad request is answered. If there is one, the value of true for the is_default field is automatically followed by the next one.

### Delete

If you tried to delete the default tenantUser and there was no other entity available to mark as default, the previous behavior was to delete the user and then delete the tenantUser entity.

Since the deletion method is configured in a cascade form, this logic could result in trying to delete the tenantUser and this entity would no longer exist, leading to errors.

I changed this behavior to first delete the tenantUser entity and then the user.

### Tests

In addition, a description of this process was added to the documentation and a unit test was added that, among other cases, also tests this possibility.

critBus committed:
fix update tenant user

critBus committed:
fix update tenant user

critBus committed:
extra doc on the default update tenant user

critBus committed:
extra doc on the default update tenant user

raulodev committed:
remove unnecesary else

critBus committed:
fix delete tenant user

If you tried to delete the default tenantUser and there was no other entity available to mark as default, the previous behavior was to delete the user and then delete the tenantUser entity.

Since the deletion method is configured in a cascade form, this logic could result in trying to delete the tenantUser and this entity would no longer exist, leading to errors.

I changed this behavior to first delete the tenantUser entity and then the user.

In addition, a description of this process was added to the documentation and a unit test was added that, among other cases, also tests this possibility.

critBus committed:
fix delete tenant user

If you tried to delete the default tenantUser and there was no other entity available to mark as default, the previous behavior was to delete the user and then delete the tenantUser entity.

Since the deletion method is configured in a cascade form, this logic could result in trying to delete the tenantUser and this entity would no longer exist, leading to errors.

I changed this behavior to first delete the tenantUser entity and then the user.

In addition, a description of this process was added to the documentation and a unit test was added that, among other cases, also tests this possibility.

raulodev committed:
create contract form field

raulodev committed:
fix create contract form fields

critBus committed:
Overview of Loguin using the swagger page

critBus committed:
Overview of Loguin using the swagger page

raulodev committed:
get contract form

raulodev committed:
update urls.py

raulodev committed:
create contract form validators

raulodev committed:
add ContractFormFieldResponse model

raulodev committed:
update admin site

raulodev committed:
update minio  utils

raulodev committed:
create contract form response

raulodev committed:
update form serializer.py

raulodev committed:
update minio client

raulodev committed:
update inspections validators.py

raulodev committed:
remove print

raulodev committed:
few changes

raulodev committed:
fix clone contract form template

raulodev committed:
remove coment

raulodev committed:
run makemigrations

raulodev commented on rental/tenantUser/api.py:
```diff
@@ -178,22 +172,30 @@ def put(self, request, tenantUser_id):
+            serialized_tenant_user = TenantUserListSerializer(updated_tenant_user)
+            return Response(serialized_tenant_user.data, status=status.HTTP_200_OK)
+        except ValidationError as ex:
+            raise BadRequest400APIException(str(ex.message))
```
Creo que hay demasidao codigo en manejar las excepciones

raulodev commented on rental/tenantUser/serializer.py:
```diff
@@ -45,15 +45,6 @@ def validate_role(self, value):
-            raise serializers.ValidationError(
-                "Invalid role. Choices are 'Admin', 'Staff', or 'Owner'."
-            )
-        return value
```
esta validacion es innecesaria  en un `ModelSerializer`  ya que drf lo hace de automaticamente por los campos del TenantUser

raulodev commented on rental/tenantUser/api.py:
```diff
@@ -178,22 +172,30 @@ def put(self, request, tenantUser_id):
+            raise ex
+        except Exception as ex:
+            raise InternalServerError500APIException()
     @extend_schema(
```
muchos  try-except

critBus replied:
El modelo TenantUser en su método ‘save’ tiene un comportamiento que responde con un ValidationError, el cual anteriormente no estaba siendo atrapado, la manera actual garantiza que lo sea.

Además el método  update_tenantUser arroja excepciones NotFound404APIException y BadRequest400APIException
 
Para atrapar cualquier error desconocido evitando que el programa explote descontroladamente  y a su vez responder correctamente ante los errores si controlados esperados, encierro el código correspondiente al update y su respuesta

critBus replied:
En el commit 534ab5465107dfcee11fb9b65e62a7c7ceae4cb0 se agrupan las excepciones controladas, pero dejo margen para atrapar un futuro ‘ValidationError’ o algun error no esperado de cualquier tipo

raulodev commented on tests/rental/tenant_user/test_delete_tenant_user.py:
```diff
@@ -0,0 +1,151 @@
+from rental.tenantUser.models import TenantUser
+
+from ..auth.parent_case.auth_api_test_case import AuthAPITestCase
+from ..auth.utils.custom_test_user import CustomTestUser
```
No uses el .. para importar

raulodev committed:
Merge pull request #18 from Nitza-Developement/contract-from

Contract from

critBus committed:
except except  in update tenant user

critBus committed:
except except  in update tenant user

critBus committed:
add doc and test create tenant

critBus committed:
add doc and test create tenant

critBus committed:
add test update tenant

remove imports relative in tests

critBus committed:
add test update tenant

remove imports relative in tests

critBus committed:
add doc update tenant

critBus committed:
add doc update tenant

critBus committed:
add test retrieve tenant

critBus committed:
add test retrieve tenant

critBus committed:
add doc retrieve tenant

critBus committed:
add doc retrieve tenant

critBus committed:
add test list pagination tenant

critBus committed:
add test list pagination tenant

critBus committed:
add test list query params tenant

critBus committed:
add test list query params tenant

critBus committed:
add doc list  tenant

critBus committed:
add doc list  tenant

critBus committed:
add test delete tenant

critBus committed:
add test delete tenant

critBus committed:
add doc delete tenant

critBus committed:
add doc delete tenant

raulodev commented on rental/tenantUser/features.py:
```diff
@@ -38,35 +41,60 @@ def get_tenantUsers(user_requesting: User):
+        if not tenant_user:
+            raise NotFound404APIException(
+                f"TenantUser with ID {tenant_user_id} doesnt exists"
+            )
```
esto no hace falta ya que cuando se hace esto  `tenant_user = get_tenantUser(tenant_user_id)` se levanta la excepcion `NotFound404APIException` en caso de no existir

critBus replied:
quite este código en el commit 3cb531abcebadbc78e6f745641035e4534329fe7

critBus committed:
remove unnecessary 404 in crud tenant user

critBus committed:
remove unnecessary 404 in crud tenant user

critBus committed:
add test and dodc create vehicle

critBus committed:
add test and dodc create vehicle

critBus committed:
add test and doc update vehicle

El metodo update_vehicle no retornaba la instancia necesaria para el serilizador en la respuesta, esto fue solucionado

La validacion personalizada con respecto al campo ‘vin’ del serilizador VehicleUpdateSerializer necesitaba una instancia para funcionar del modelo Vehicle que no se le estaba proveyendo, esto se  soluciono agregándole al principio del metodo put del endpoint una obtención y comprobación de la instancia a editar, que posteriormente se le paso al serializador para que este funcionase

Al realizar este cambio, tambien se utilizó esta instancia en la validación personalizada del campo ‘plate’

Como se tenia una instancia de Vehicle, se adapto el método update_vehicle para que la utilizase y con esto remover la comprobación de existencia de la misma que para este punto ya seria innecesaria

critBus committed:
add test and doc update vehicle

El metodo update_vehicle no retornaba la instancia necesaria para el serilizador en la respuesta, esto fue solucionado

La validacion personalizada con respecto al campo ‘vin’ del serilizador VehicleUpdateSerializer necesitaba una instancia para funcionar del modelo Vehicle que no se le estaba proveyendo, esto se  soluciono agregándole al principio del metodo put del endpoint una obtención y comprobación de la instancia a editar, que posteriormente se le paso al serializador para que este funcionase

Al realizar este cambio, tambien se utilizó esta instancia en la validación personalizada del campo ‘plate’

Como se tenia una instancia de Vehicle, se adapto el método update_vehicle para que la utilizase y con esto remover la comprobación de existencia de la misma que para este punto ya seria innecesaria

critBus committed:
add test and doc list vehicle

critBus committed:
add test and doc list vehicle

critBus committed:
add test delete vehicle

critBus committed:
add test delete vehicle

critBus committed:
Merge branch 'add-basic-swagger' of https://github.com/Nitza-Developement/rental_backend into add-basic-swagger
